prestashopv1/plugin/modules/indodana/controllers/front/validation.php

```diff
@@ -0,0 +1,158 @@
+            $message = null;
+        } else {
+            $paymentStatus = Configuration::get('PS_OS_ERROR');
+
+            /**
+             * Add a message to explain why the order has not been validated
+             */
+            $message = $this->module->l('An error occurred while processing payment');
+        }
+
+        $this->module->validateOrder(
+            $cart->id,
+            $paymentStatus,
+            $total,
+            $this->module->displayName,
+            $message,
+            [],
+            (int) $currency->id,
+            false,
+            $customer->secure_key
+        );
+
+        $orderId = Order::getOrderByCartId((int) $cart->id);
+        $orderDetails = new Order((int) $orderId);
+
+        /**
+         * An error occured and is shown on a new page.
+         */
+        if (!$orderId) {
+            $this->errors[] = $this->module->l('An error occured. Please contact the merchant to have more informations');
+
+            return $this->setTemplate('error.tpl');
+        }
+
+
+        /**
+         * The order has been placed so we redirect the customer on the indodana payment page.
+         */
+        $moduleId = $this->module->id;
+        $approveUrl = $this->context->link->getModuleLink(
+            $this->module->name,
+            'approve',
+            ['id_order' => $orderId],
+            true
+        );
+        $cancelUrl = $this->context->link->getModuleLink(
+            $this->module->name,
+            'cancel',
+            ['id_order' => $orderId],
+            true
+        );
+        $lang = $this->context->language->iso_code;
+        $backUrl = _PS_BASE_URL_ . '/' . $lang  . '/order-confirmation?id_cart=' . $cart->id . '&id_module=' . $moduleId . '&id_order=' . $orderId . '&key=' . $customer->secure_key;
+
+        $checkoutUrl = IndodanaTools::getIndodanaCommon()->checkout([
+            'merchantOrderId' => $orderId . '-' . $orderDetails->reference,
+            'totalAmount' => IndodanaTools::getTotal($cart),
+            'discountAmount' => IndodanaTools::getDiscount($cart),
+            'shippingAmount' => IndodanaTools::getShippingFee($cart),
+            'taxAmount' => IndodanaTools::getTax($cart),
+            'products' => IndodanaTools::getProducts($cart),
+            'customerDetails' => IndodanaTools::getCustomerDetails($orderId),
+            'billingAddress' => IndodanaTools::getBillingAddress($orderId),
+            'shippingAddress' => IndodanaTools::getDeliveryAddress($orderId),
+            'paymentType' => $_POST['indodana_selection'],
+            'approvedNotificationUrl' => $approveUrl,
+            'cancellationRedirectUrl' => $cancelUrl,
+            'backToStoreUrl' => $backUrl
+        ]);
+
+        Tools::redirect($checkoutUrl);
+    }
+
+    protected function isValidOrder()
+    {
+        /*
+         * Add your checks right there
+         */
+        return true;
```

Comment:
Do you need to confirm anything here to us?

Reply:
No, this method is generated by Prestashop generator, guess I'll remove this method